src/GafferSceneUI/LightTool.cpp

```diff
@@ -0,0 +1,1841 @@
+				// to get smaller / larger, quickly going to zero / 180.
+				// We can avoid the feedback loop by using raster coordinates, which unproject
+				// the local coordinates to a fixed frame of reference (the screen).
+				const Line3f dragLine( event.line.p0, event.line.p1 );
+
+				newAngle = radiansToDegrees(
+					atan2( rasterDragDistance( dragLine ) + m_rasterXOffset, m_rasterZPosition )
+				);
+			}
+			else if( m_drag.value().isLinearDrag() )
+			{
+				// Intersect the gadget-local `event` line with the sphere centered at the gadget
+				// origin with radius equal to the distance along the handle where the user clicked.
+				// `Imath::Sphere3::intersect()` returns the closest (if any) intersection, but we
+				// want the intersection closest to the handle line, so we do the calculation here.
+
+				const Line3f eventLine( event.line.p0, event.line.p1 );
+
+				const float B = 2.f * ( eventLine.dir ^ eventLine.pos );
+				const float C = ( eventLine.pos ^ eventLine.pos ) - ( m_arcRadius * m_arcRadius );
+
+				const float discriminant = B * B - 4.f * C;
+				// If discriminant is negative, the click is outside the sphere.
+				if( discriminant < 0 )
+				{
+					return true;
+				}
+				else
+				{
+					// t = ( -B +/- sqrt( B^2 - 4AC ) ) / 2A ( A = 1 )
+					const float sqRoot = std::sqrt( discriminant );
+
+					const V3f minusP = eventLine( ( -B - sqRoot ) * 0.5f );
+					const V3f plusP = eventLine( ( -B + sqRoot ) * 0.5f );
+
+					if( minusP.z > 0 && plusP.z > 0 )
+					{
+						newAngle = 180.f;
+					}
+					else if( minusP.z >= 0 && plusP.z < 0 )
+					{
+						newAngle = radiansToDegrees( atan2( -plusP.x, -plusP.z ) );
+					}
+					else if( minusP.z < 0 && plusP.z >= 0 )
+					{
+						newAngle = radiansToDegrees( atan2( -minusP.x, -minusP.z ) );
+					}
+					else
+					{
+						const auto &[coneAngle, penumbraAngle] = spotLightAngleValues();
+						const float angle = handleAngle( m_handleType == HandleType::Cone ? coneAngle : penumbraAngle.value() );
+						const M44f r = M44f().rotate( V3f( 0, degreesToRadians( angle ), 0 ) );
+						const Line3f handleLine( V3f( 0 ), V3f( 0, 0, -m_arcRadius ) * r );
+
+						const V3f p = handleLine.distanceTo( minusP ) < handleLine.distanceTo( plusP ) ? minusP : plusP;
+						newAngle = radiansToDegrees( atan2( -p.x, -p.z ) );
+					}
+				}
+			}
+			else
+			{
+				// All other drags can use the `AngularDrag` directly.
+				newAngle = radiansToDegrees( m_drag.value().updatedRotation( event ) );
+			}
+
+			float angleScale = m_dragStartAngle > 0 ? newAngle / m_dragStartAngle : newAngle;
+
+			// Find an `angleScale` that all lights can use without going past limits for
+			// penumbra and cone angles. Cone angles can go to zero, requiring the check
+			// against divide-by-zero. Penumbra handles are hidden close to zero, so they
+			// don't need to be checked.
+
+			for( auto &[coneInspection, originalConeAngle, penumbraInspection, originalPenumbraAngle] : m_inspections )
```

Comment:
I really don't much like this loop. It seems the logic it's implementing is legitimately quite complicated, and some messiness is probably required here, but the current structure leaves me with a bunch of questions:
* is angleScale ever < 0? Looking through all the cases, it seems like the times when we modify the second value in max, it is to raise it above 0? If that's true, then we could put one max with 0 at the end, and remove it from inside most branches.
* is there ever a situation where the second value that is passed to min on one loop iteration is less than the second value passed to max on another iteration? Then the initial value of angleScale would be ignored, and the resulting value would depend solely on which order m_inspections are visited in, which seems bad. It's very likely that this can't happen - but I don't see any easy way to prove it can't by quickly inspecting this code.
* It may be a bit late for any reconsideration of user behaviour, if this has already been validated, but I do wonder whether this is the simplest and most desirable approach. First off, to confirm I'm understanding what this is currently intended to do in the simplest case: suppose I have light A with a cone angle of 60, and light B with a cone angle of 120, and I'm using a penumbra handle with absolute mode, and both penumbra's are currently set the same as the cone angle. Because light B has a large cone angle, the maximum angleScale permitted is `180/120 = 1.5`. This means that if editing light A at the same time, if I drag A's penumbra larger, it will be locked to a maximum of `60 * 1.5 = 90` degrees because of the limitation on B, right? Is this actually desirable? The more natural approach to me would seem to be to not clamp angleScale, but for each selected target, apply angleScale, then apply appropriate clamping ... it seems like the logic would be a lot simpler if just clamping one at a time, and I'm not sure what the advantage is of clamping out light A because light B has hit its maximum.

Reply:
I'm not 100% convinced this is a better behavior, but for consideration it's in https://github.com/GafferHQ/gaffer/pull/5218/commits/a8fb0e07f5e018555f665db99bd48a9be81cdff9.

My hesitation is in the case of :
1. Two lights in the same scene, one with a penumbra value that is pretty close to the cone angle (so not far for it to go before it's clamped).
2. Drag the cone handle whose penumbra is almost clamped, it will move a few degrees and then stop as expected.
3. The other light is still moving because it hasn't been clamped, but there's no feedback in the UI - the light with the handles is not changing, and neither is the arc indicator for the angles.

I'm not dead-set against independent clamping, especially with the elimination of the scaling behavior of the values this allows, but that's my main point against it at the moment.

Reply:
Hmm, that's a fair concern. The code does seem a whole lot clearer this way, but it looks like maybe part of that is a refactor that could be kept even if the behaviour went back to the old behaviour?

I suppose what your point suggests is clamping based on the light with the handles, and the light being edited, but maybe not based on all selected lights? i.e. in your situation 2, both lights would clamp after the handle reaches maximum, but if you dragged the handle on the other light instead, then you could keep dragging larger even after the almost-clamped penumbra has started clamping?

For any of these approaches, maybe we need to to package up clamping into a function that can perform clamping for an m_inspection, and can be called wherever we decide it's needed?

Reply:
I like the idea of clamping based on either the active light or each light individually, whichever comes first. I implemented that in https://github.com/GafferHQ/gaffer/pull/5218/commits/416630b7210cf40226374f9fe9716279a47a84b8.

It was a bit of a rabbit hole because I wasn't holding onto _both_ the cone and penumbra data for the active light, both of which are needed to figure out the clamping. But I think my solution also improved the general data handling, and forced me to be more explicit about when cone angles are for the plug and for the handle (the difference between what the renderer expects and what the handle represents).

Settling on manipulating handle-space values until the last minute setting of a plug allowed me to remove all of the `handleAngle` calls too.

Finally, I extracted the clamp function to its own method so it's reusable.

All of that is in the final commit, https://github.com/GafferHQ/gaffer/pull/5218/commits/416630b7210cf40226374f9fe9716279a47a84b8, passing my use-tests and I think feeling pretty good!

Reply:
This all looks quite good to me, and I think the behaviour makes a lot of sense, hopefully John agrees.

There is one minor stylistic thing, which is minor enough that I wouldn't recommend changing it here, but for future reference: in this last commit, you've switched originalPenumbraAngleData ( which is just used as an input to handleAngle() ) from float to FloatDataPtr - and it appears you don't need any of the runtime polymorphism of Data, it's just to have an optional value? I think I would prefer std::optional here ( the performance isn't important, but it documents intent more clearly ), especially since the output from handleAngle() is an std::optional - but it's a slight preference.

Reply:
> This all looks quite good to me, and I think the behaviour makes a lot of sense, hopefully John agrees.

The latest clamping behaviour you both described sounds good to me.